pkg/rbac/ensure/ensurer.go

```diff
@@ -0,0 +1,81 @@
+package ensure
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/rancher/wrangler/pkg/generic"
+)
+
+// Ensurer is used to ensure that a group of objects has been created in k8s, and that all related corrupted versions
+// of the object are properly processed.
+type Ensurer[T generic.RuntimeMetaObject] struct {
+	// ExistingStrategy is the strategy for retrieving existing objects that should be compared with the requested
+	// objects
+	ExistingStrategy ExistingStrategy[T]
+	// ObjectIdentifier produces a string that is unique to T. Used to tell if a requested object
+	// is covered by exising objects
+	ObjectIdentifier func(T) string
+	// IsValid is used to determine if a current object should be processed by ValidStrategy
+	// or by InvalidStrategy
+	IsValid func(obj T) bool
+	// InvalidStrategy describes how to process invalid objects. Invalid objects will not be matched with requested objects
+	// after processing, and are not altered further by Ensure.
+	InvalidStrategy InvalidObjStrategy[T]
+	// ValidStrategy describes how to process valid objects. Valid objects are objects which not currently covered by
+	// an existing object, but were requested in parameters of Ensure
+	ValidStrategy ValidObjStrategy[T]
```

Comment:
I think this might be better as 
```suggestion
	MissingStrategy MissingObjStrategy[T]
```
My thinking is that this isn't a strategy applied to valid objects, but instead for objects that currently don't exist that we want.